### ISSUE TITLE
run -P: TOCTOU in port allocation

### ISSUE THREAD

**Issue by edsantiago** [containers/podman#19048]
The local port allocated with `run -P` seems unsafe when multiple jobs start at once:
```
$ podman [options] run --name portcheck -dt -P --health-cmd curl -f http://localhost/ quay.io/libpod/alpine_nginx:latest
Error: rootlessport listen tcp 0.0.0.0:43185: bind: address already in use
```
And it's not just rootless:
```
# podman [options] run -dt -P --health-cmd curl -f http://localhost/ quay.io/libpod/alpine_nginx:latest
Error: cannot listen on the TCP port: listen tcp4 :42611: bind: address already in use
```
In both cases the ports had been grabbed by a different `nginx` test.

From my look last night I think this is a podman/rootlessport bug, not something that can be fixed in e2e tests. So even though this is infrequent, it's something that can probably manifest in the field.

* fedora-38 : int podman fedora-38 root host boltdb
    * [05-18-2023 14:08](https://api.cirrus-ci.com/v1/artifact/task/6453053784588288/html/int-podman-fedora-38-root-host-boltdb.log.html#t--podman-container-port-l-nginx--1) in podman container port  -l nginx
* fedora-38 : int podman fedora-38 rootless host sqlite
    * [06-28 18:38](https://api.cirrus-ci.com/v1/artifact/task/4611663630958592/html/int-podman-fedora-38-rootless-host-sqlite.log.html#t--Podman-port-podman-port-nginx-by-name--1) in Podman port podman port nginx by name

**Comment by Luap99** [comment]
You are correct this is a race within podman. The current flow is allocate the ports once on container creation which obviously leads to race conditions with other processes. It is a long standing issue: https://github.com/containers/podman/issues/10205#issuecomment-1010055023

**Comment by purunjay-bhardwaj** [comment]
Traced this through the code. `utils.GetRandomPort()` (`utils/ports.go`) probes a free port via `net.Listen("tcp",":0")` then immediately closes it — this happens at container-create time in `pkg/specgen/generate/ports.go`. The actual bind doesn't happen until later, in a different process/netns, via `cmd/rootlessport/main.go`'s `exposePorts`→`validateAndAddPort`→`pm.AddPort`, which has no retry on `EADDRINUSE`. Any process (including another concurrent `podman run -P`) that grabs the port in that window causes the hard failure seen here.

A full fix likely needs either holding the probe socket open until the real bind (hard — crosses a process/netns boundary), or a bounded retry-with-fresh-port loop when the original request was for a random (not user-pinned) port. Happy to attempt the retry-loop approach if that's the preferred direction — wanted to confirm before proposing a specific diff given how central this path is.

**Comment by Honny1** [comment]
PTAL @Luap99

**Comment by Luap99** [comment]
A retry loop cannot work, if the port is blocked it is blocked we should not waste time hoping it will be free again.

This is a rather complex issue like I described in https://github.com/podman-container-tools/podman/issues/10205#issuecomment-1010055023 The port must be bound exactly once with port 0 at container start time to get a random free and and then use that as long ass the container is running. I don't see us being ready for that anytime soon, if ever. As root where we bond ports in libpod that might be an option but even that is quite complex and involved for very marginal benifits.

We still have 5 different ways how binds get bound so doing this right is not feasible at the moment.
